From a09a307df96f91731f91eab4a6c1b0af9ddd24c2 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 31 Mar 2022 20:07:08 +0300
Subject: [PATCH] [SW-77251] gaudi3: add valid address check api

This patch add valid address check api for gaudi3 which
will be used by fit operations.

Change-Id: Icd219260ea5834fc1de82ea9ce3c38182cc1e56f
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 592c246..81b7554 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -666,3 +666,34 @@
 		execute_recovery();
 	}
 }
+
+bool arcboot_is_valid_address_range(uint64_t addr, int len)
+{
+	if (addr >= SP_SRAM_BASE &&
+	    addr + len <= SP_SRAM_BASE + HL_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_SP_SRAM_BASE &&
+	    addr + len <= HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_ARC0_ICCM_FULL_BASE &&
+	    addr + len <= HL_ARC0_ICCM_FULL_BASE + ARC0_MEM_SIZE)
+		return true;
+
+	if (addr >= HL_ARC1_ICCM_FULL_BASE &&
+	    addr + len <= HL_ARC1_ICCM_FULL_BASE + ARC1_MEM_SIZE)
+		return true;
+
+	if (addr >= HL_ARC2_ICCM_FULL_BASE &&
+	    addr + len <= HL_ARC2_ICCM_FULL_BASE + ARC2_MEM_SIZE)
+		return true;
+
+	if (addr >= HL_SRAM_BASE && addr + len <= HL_SRAM_BASE + HL_SRAM_SIZE)
+		return true;
+
+	if (addr >= HL_SRAM_BASE_FULL && addr + len <= HL_SRAM_BASE_FULL + HL_SRAM_SIZE)
+		return true;
+
+	return false;
+}
diff --git a/arch/gaudi3/arc/soc.h b/arch/gaudi3/arc/soc.h
index f5071d8..63ff406 100644
--- a/arch/gaudi3/arc/soc.h
+++ b/arch/gaudi3/arc/soc.h
@@ -40,6 +40,10 @@
 #define SPI_FLASH_BLOCK_SIZE		1024
 
 #define REGION_BASE_ADDR(REG_IDX)	(REG_IDX << 28)
+#define SZ_32K				0x00008000
+#define SZ_64K				0x00010000
+#define SZ_128K				0x00020000
+#define SZ_256K				0x00040000
 #define SZ_256MB			0x10000000
 /* 2MB for Gaudi2 . TODO: Confirm for gaudi3 */
 #define FLASH_VALID_MASK		0x1FFFFF
@@ -83,6 +87,7 @@
 #define HL_SCRATCHPAD_SRAM_OFFSET	0x3120000
 #define HL_FWSRAM_OFFSET		0x3140000
 #define HL_PCIE_FWSRAM_OFFSET		0x7F00000
+#define SP_SRAM_BASE			(HL_GEN_MEM_BASE + HL_SCRATCHPAD_SRAM_OFFSET)
 
 #define HL_FW_MEM_BASE_FULL		0x0300007FD0000000ull
 #define HL_GEN_MEM_BASE_FULL		HL_FW_MEM_BASE_FULL
@@ -90,5 +95,18 @@
 #define HL_SP_SRAM_SIZE			0x10000
 #define HL_MIRROR_CFG_BASE		HL_D2D_LBW_BASE
 
+#define HL_ARC0_ICCM_FULL_BASE		(HL_FW_MEM_BASE_FULL + HL_ARC0_ICCM_OFFSET)
+#define ARC0_ICCM_SIZE			SZ_128K
+#define ARC0_DCCM_SIZE			SZ_32K
+#define HL_ARC1_ICCM_FULL_BASE		(HL_FW_MEM_BASE_FULL + HL_ARC1_ICCM_OFFSET)
+#define ARC1_ICCM_SIZE			SZ_256K
+#define ARC1_DCCM_SIZE			SZ_64K
+#define HL_ARC2_ICCM_FULL_BASE		(HL_FW_MEM_BASE_FULL + HL_ARC2_ICCM_OFFSET)
+#define ARC2_ICCM_SIZE			SZ_128K
+#define ARC2_DCCM_SIZE			SZ_32K
+#define ARC0_MEM_SIZE			(ARC0_ICCM_SIZE + ARC0_DCCM_SIZE)
+#define ARC1_MEM_SIZE			(ARC1_ICCM_SIZE + ARC1_DCCM_SIZE)
+#define ARC2_MEM_SIZE			(ARC2_ICCM_SIZE + ARC2_DCCM_SIZE)
+
 #define CONV_64B_TO_ARC(addr)	((addr & 0xFFFFFFF) + HL_FW_MEM_BASE)
 #endif /* _SOC_H_ */
diff --git a/include/arcboot.h b/include/arcboot.h
index c4ca14c..6ae44e9 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -221,6 +221,7 @@
 void arcboot_rom_load_recovery(void);
 int arcboot_rom_mem_init(uint64_t addr, uint32_t size);
 void arcboot_rom_verify(void);
+bool arcboot_is_valid_address_range(uint64_t addr, int len);
 uint32_t arcboot_error_read_and_clr(uint32_t *major, uint32_t *minor);
 bool arcboot_error_inform(uint32_t major, uint32_t minor);
 bool arcboot_error_is_pending(void);
